rasa_sdk/knowledge_base/actions.py

```diff
@@ -131,22 +131,37 @@ async def run(
         object_type = tracker.get_slot(SLOT_OBJECT_TYPE)
         last_object_type = tracker.get_slot(SLOT_LAST_OBJECT_TYPE)
         attribute = tracker.get_slot(SLOT_ATTRIBUTE)
+        has_mention = tracker.get_slot(SLOT_MENTION) is not None
 
-        new_request = object_type != last_object_type
+        # check if attribute entity is found in latest user message. This is used
+        # to track whether the request is to query objects or query attributes
+        has_attribute_in_latest = any(
+            entity.get("entity") == "attribute"
+            for entity in tracker.latest_message["entities"]
+        )
 
         if not object_type:
-            # object type always needs to be set as this is needed to query the
-            # knowledge base
-            dispatcher.utter_message(response="utter_ask_rephrase")
-            return []
-
-        if not attribute or new_request:
+            # sets the object type dynamically from entities if object_type is not
+            # found in user query
+            object_types = self.knowledge_base.get_object_types()
+            object_type = match_extracted_entities_to_object_type(tracker, object_types)
+            set_object_type_slot_event = [SlotSet(SLOT_OBJECT_TYPE, object_type)]
+            tracker.add_slots(
+                set_object_type_slot_event
+            )  # temporarily set the `object_type_slot` to extracted value
+
+        if object_type and not has_attribute_in_latest:
             return await self._query_objects(dispatcher, object_type, tracker)
-        elif attribute:
+        elif object_type and attribute:
```

Comment:
What about `elif object_type and has_attribute_in_latest`? Is this equivalent to this branch `elif object_type and attribute`?

Reply:
Functionally, both could be equivalent. But we wanted to preserve the `elif attribute` condition used in Rasa’s implementation. We wanted to keep it that way due to the following reasons; 
1. One of the input parameters of the `query_attribute()` method is the attribute value in `SLOT_ATTRIBUTE`. The `elif object_type and attribute` condition directly checks if the attribute slot is set. This check can also come in handy in the instances where the `SLOT_ATTRIBUTE` is not reset to `None`. 
2. Inside `query_attribute()` there are `if` conditions to check for the attribute value.  By checking this in `elif object_type and attribute` we could ensure that these conditions are running as expected.
3. Additionally, In some cases the `SLOT_ATTRIBUTE` value is not reset to None. Therefore, we cannot use `if object_type and not attribute` as a condition to list objects. Which is why we had to use the `if object_type and not has_attribute_in_latest` condition to list objects. 

However, if this is not a concern, we could change the condition from `elif object_type and attribute` to `elif object_type and has_attribute_in_latest_message`

Reply:
All clear, thanks for the explanation, we can keep the existing `elif object_type and attribute`.